test/core/security/credentials_test.cc

```diff
@@ -1512,16 +1512,105 @@ static void test_no_google_default_creds(void) {
+      reinterpret_cast<grpc_composite_channel_credentials*>(
+          grpc_google_default_credentials_create(call_creds));
+  GPR_ASSERT(g_test_gce_tenancy_checker_called == true);
+  GPR_ASSERT(channel_creds != nullptr);
+  GPR_ASSERT(channel_creds->call_creds() != nullptr);
+  grpc_httpcli_set_override(compute_engine_httpcli_get_success_override,
+                            httpcli_post_should_not_be_called);
+  run_request_metadata_test(channel_creds->mutable_call_creds(), auth_md_ctx,
+                            state);
+  grpc_core::ExecCtx::Get()->Flush();
+  channel_creds->Unref();
+  grpc_httpcli_set_override(nullptr, nullptr);
+}
+
+struct fake_call_creds : public grpc_call_credentials {
+ public:
+  explicit fake_call_creds() : grpc_call_credentials("fake") {
+    grpc_slice key = grpc_slice_from_static_string("foo");
+    grpc_slice value = grpc_slice_from_static_string("oof");
+    dummy_md_ = grpc_mdelem_from_slices(key, value);
+    grpc_slice_unref(key);
+    grpc_slice_unref(value);
+  }
+
+  ~fake_call_creds() { GRPC_MDELEM_UNREF(dummy_md_); }
+
+  bool get_request_metadata(grpc_polling_entity* pollent,
+                            grpc_auth_metadata_context context,
+                            grpc_credentials_mdelem_array* md_array,
+                            grpc_closure* on_request_metadata,
+                            grpc_error** error) {
+    grpc_credentials_mdelem_array_add(md_array, dummy_md_);
+    return true;
+  }
+
+  void cancel_get_request_metadata(grpc_credentials_mdelem_array* md_array,
+                                   grpc_error* error) {}
+
+ private:
+  grpc_mdelem dummy_md_;
+};
+
+static void test_google_default_creds_not_default(void) {
+  expected_md emd[] = {{"foo", "oof"}};
+  request_metadata_state* state =
+      make_request_metadata_state(GRPC_ERROR_NONE, emd, GPR_ARRAY_SIZE(emd));
+  grpc_auth_metadata_context auth_md_ctx = {test_service_url, test_method,
+                                            nullptr, nullptr};
+  grpc_core::ExecCtx exec_ctx;
+  grpc_flush_cached_google_default_credentials();
+  grpc_core::RefCountedPtr<grpc_call_credentials> call_creds =
+      grpc_core::MakeRefCounted<fake_call_creds>();
+  set_gce_tenancy_checker_for_testing(test_gce_tenancy_checker);
+  g_test_gce_tenancy_checker_called = false;
+  g_test_is_on_gce = true;
+  grpc_httpcli_set_override(
+      default_creds_metadata_server_detection_httpcli_get_success_override,
+      httpcli_post_should_not_be_called);
+  grpc_composite_channel_credentials* channel_creds =
+      reinterpret_cast<grpc_composite_channel_credentials*>(
+          grpc_google_default_credentials_create(call_creds.release()));
+  GPR_ASSERT(g_test_gce_tenancy_checker_called == true);
+  GPR_ASSERT(channel_creds != nullptr);
+  GPR_ASSERT(channel_creds->call_creds() != nullptr);
+  run_request_metadata_test(channel_creds->mutable_call_creds(), auth_md_ctx,
+                            state);
```

Comment:
Should we expect the RPC to fail here, since the call credentials are not google default credentials?

Reply:
This test doesn't actually ever send an RPC. It just runs the metadata call credentials code path, so we can't make such an expectation. Setting up a test that actually sent an RPC like this would require a hermetic server that accepts a Google default credentials, which AFAIK, has never been done.

Reply:
> If used with any other sort of call credential, the connection may suddenly and unexpectedly begin failing RPCs.`

I wanted to do a sanity check here, might be wrong. This tests seems not failing the call credentials and still invokes it (metadata changed). Is the statement above tested?

Reply:
This is not a statement of expected behavior of the function, but rather, of explicitly *undefined behavior*. The key word is "_may_". We would much rather reject a call credential that does not implement the Application Default Credentials mechanism; the problem is that this mechanism is defined by a convention, rather than by any C++ or Python type.

If we took a dependency on the auth library instead of the other way around, then we could identify their implementation of the Application Default Credentials mechanism and allow it and only it, failing any other type of call credential. Unfortunately, since the dependency is in the other direction, we have no way of knowing *which* call credential that might be. So, instead, we opted to add this warning to the docstring for the Core API.

Reply:
I misunderstood the docstring. I think we can promise less here. I guess we can emphasize that using other type of call credentials just results in undefined behavior?

Reply:
Rephrased so that it does not use the word "may."